### Analyzing Air Quality in Python
### Interesting question: What states in the Unites States have the [worst air quality and pollution](https://www.lung.org/research/sota/city-rankings/most-polluted-cities) and how much has it changed in the last couple of years?

To test this question, we will start with the the OpenWeather API. 

### Python Requests Library Method (connecting directly to the API)

In [1]:
import requests

In [2]:
from appid import weather_api_key

We can find the current air pollution data for Los Angeles, California using its latitude and longitude coordinates.

In order to call the current air pollution data, we need to follow a few steps:

Separate out the params (lat, lon, units, appid)

In [3]:
params = {'lat': '34.0522',
          'lon': '118.2437',
          'units': 'imperial',
          'appid': weather_api_key}
response = requests.get('https://api.openweathermap.org/data/2.5/air_pollution', params=params)

In [4]:
response.json()

{'coord': {'lon': 118.2437, 'lat': 34.0522},
 'list': [{'main': {'aqi': 2},
   'components': {'co': 273.71,
    'no': 0,
    'no2': 3.3,
    'o3': 74.39,
    'so2': 2.86,
    'pm2_5': 16.11,
    'pm10': 16.54,
    'nh3': 0.77},
   'dt': 1687039559}]}

As seen above, we just called the current air pollution data for Los Angeles, CA. According to the American Lung Association, it is the city with the lowest air quality by ozone. Here we know (among other things):

The Air Quality Index ('aqi'): Possible values: 1, 2, 3, 4, 5. Where 1 = Good, 2 = Fair, 3 = Moderate, 4 = Poor, 5 = Very Poor
The concentration of:
CO (Carbon monoxide), μg/m3
NO (Nitrogen monoxide), μg/m3
NO2 (Nitrogen dioxide), μg/m3
O3 (Ozone), μg/m3
SO2 (Sulphur dioxide), μg/m3
PM2.5 (Fine particles matter), μg/m3
PM10 (Coarse particulate matter), μg/m3
NH3 (Ammonia), μg/m3

In [5]:
params = {'lat': '35.3733',
          'lon': '119.0187',
          'units': 'imperial',
          'appid': weather_api_key}
response = requests.get('https://api.openweathermap.org/data/2.5/air_pollution', params=params)

In [6]:
response.json()

{'coord': {'lon': 119.0187, 'lat': 35.3733},
 'list': [{'main': {'aqi': 3},
   'components': {'co': 270.37,
    'no': 0,
    'no2': 2.74,
    'o3': 75.82,
    'so2': 2.53,
    'pm2_5': 34.07,
    'pm10': 38.06,
    'nh3': 0.37},
   'dt': 1687039559}]}

As seen above, we just called the current air pollution data for Bakersfield, CA, which is the second most polluted. 

In [7]:
params = {'lat': '36.3302',
          'lon': '119.2921',
          'units': 'imperial',
          'appid': weather_api_key}
response = requests.get('https://api.openweathermap.org/data/2.5/air_pollution', params=params)

In [8]:
response.json()

{'coord': {'lon': 119.2921, 'lat': 36.3302},
 'list': [{'main': {'aqi': 3},
   'components': {'co': 277.04,
    'no': 0.02,
    'no2': 5.06,
    'o3': 63.66,
    'so2': 6.08,
    'pm2_5': 31.17,
    'pm10': 38.9,
    'nh3': 4.24},
   'dt': 1687039559}]}

Next, we called the current air pollution data for Visalia, CA, which is the third most polluted.

### Python Library Method (using a Python library built specifically for the API)

In [9]:
!pip install pyowm

In [10]:
from pyowm import OWM

Air Pollution Concentrations and Air Quality Index via Geographic Coordinates

We can also obtain the air pollution concentrations and the air quality index via the location's latitude and longitude coordinates. Here we observe the air pollution concentrations and air quality index of London, Great Britain. Specifically, we received data regarding:

CO - carbon monoxide
NO - nitrogen monoxide
NO2 - nitrogen dioxide
O3 - ozone
SO2 - sulphur dioxide
PM2_5 - fine particles matter
PM10 - coarse particulate matter
NH3 - ammonia
aqi - air quality index

In [11]:
owm = OWM(weather_api_key)

In [12]:
mgr = owm.airpollution_manager()

In [13]:
air_status = mgr.air_quality_at_coords(51.507351, -0.127758)

In [14]:
air_status.aqi 

2

In [15]:
air_status.co

240.33

In [16]:
air_status.no

0

In [17]:
air_status.no2

38.39

In [18]:
air_status.o3

66.52

In [19]:
air_status.so2

12.28

In [20]:
air_status.pm2_5

14.78

In [21]:
air_status.pm10

15.68

In [22]:
air_status.nh3

0.69

### Web Scraping

In [23]:
import pandas as pd

In [24]:
pd.read_html('https://www.airnow.gov/aqi/aqi-basics/')

[  Daily AQI Color               Levels of Concern Values of Index  \
 0           Green                            Good         0 to 50   
 1          Yellow                        Moderate       51 to 100   
 2          Orange  Unhealthy for Sensitive Groups      101 to 150   
 3             Red                       Unhealthy      151 to 200   
 4          Purple                  Very Unhealthy      201 to 300   
 5          Maroon                       Hazardous  301 and higher   
 
                           Description of Air Quality  
 0  Air quality is satisfactory, and air pollution...  
 1  Air quality is acceptable. However, there may ...  
 2  Members of sensitive groups may experience hea...  
 3  Some members of the general public may experie...  
 4  Health alert: The risk of health effects is in...  
 5  Health warning of emergency conditions: everyo...  ]

The above is helpful information regarding intepreting AQI (Air Quality Index) data

In [25]:
pd.read_html('https://www.epa.gov/air-trends/air-quality-national-summary')

[                  Unnamed: 0 1980 vs 2022 1990 vs 2022 2000 vs 2022  \
 0            Carbon Monoxide          -88          -81          -67   
 1                       Lead          ---          ---          ---   
 2  Nitrogen Dioxide (annual)          -66          -60          -52   
 3  Nitrogen Dioxide (1-hour)          -65          -54          -38   
 4             Ozone (8-hour)          -29          -22          -17   
 5             PM10 (24-hour)          ---          -34          -30   
 6             PM2.5 (annual)          ---          ---          -42   
 7            PM2.5 (24-hour)          ---          ---          -42   
 8    Sulfur Dioxide (1-hour)          -94          -90          -85   
 
    2010 vs 2022  
 0           -27  
 1           -88  
 2           -27  
 3           -21  
 4            -7  
 5            21  
 6           -21  
 7           -16  
 8           -75  ,
                          Unnamed: 0 1980 vs 2022  1990 vs 2022  2000 vs 2022  \
 0    

The above is helpful information regarding the air trends for the air quality in the United States for the last couple of decades

In [26]:
import requests
from bs4 import BeautifulSoup

In [27]:
response = requests.get('https://gispub.epa.gov/air/trendsreport/2021/#naaqs')
soup = BeautifulSoup(response.text, 'lxml')

In [28]:
response.text

'<!DOCTYPE html>\r\n<html lang="en">\r\n  <head>\r\n    <meta charset="utf-8">\r\n\t <meta http-equiv="X-UA-Compatible" content="IE=edge" />\r\n    <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">\r\n\t<meta name="description" content="Annual Air Trends Report" />\r\n    <meta name="author" content="U.S. EPA Office of Air and Radiation" />\r\n\t<meta name="keywords" content="EPA air, EPA air trends, air trends, environment,air,trends,naaqs,health,air quality,concentrations,emissions,air pollution,pollutants,EPA, toxics, air toxics" />\r\n\t<!-- Open Graph -->\r\n\t<meta property="og:title" content="Air Quality Trends Show Clean Air Progress" />\r\n\t<meta property="og:type" content="website" />\r\n\t<meta property="og:url" content="https://gispub.epa.gov/air/trendsreport/2021/" />\r\n\t<meta property="og:image" content="https://gispub.epa.gov/air/trendsreport/2021/img/AirTrends_Flyer_opengraph.png" />\r\n\t<meta property="og:description" content="R

In [29]:
soup

<!DOCTYPE html>
<html lang="en">
<head>
<meta charset="utf-8"/>
<meta content="IE=edge" http-equiv="X-UA-Compatible"/>
<meta content="width=device-width, initial-scale=1, shrink-to-fit=no" name="viewport"/>
<meta content="Annual Air Trends Report" name="description"/>
<meta content="U.S. EPA Office of Air and Radiation" name="author"/>
<meta content="EPA air, EPA air trends, air trends, environment,air,trends,naaqs,health,air quality,concentrations,emissions,air pollution,pollutants,EPA, toxics, air toxics" name="keywords"/>
<!-- Open Graph -->
<meta content="Air Quality Trends Show Clean Air Progress" property="og:title"/>
<meta content="website" property="og:type"/>
<meta content="https://gispub.epa.gov/air/trendsreport/2021/" property="og:url"/>
<meta content="https://gispub.epa.gov/air/trendsreport/2021/img/AirTrends_Flyer_opengraph.png" property="og:image"/>
<meta content="Read EPA's annual air trends report, Our Nation's Air, to learn about air quality improvements to protect hum

Here we can perform a deep dive into the air quality in the United States via data collected in the United States from 2000 to 2016. The dataset contains information on four major pollutants (Nitrogen Dioxide, Sulphur Dioxide, Carbon Monoxide, and Ozone) on a daily basis from 2000 to 2016.

In [30]:
pip install plotly

In [31]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import plotly.express as px

In [32]:
df = pd.read_csv("../Class 4/pollution.csv")

In [33]:
df.head()

Unnamed: 0  State Code  County Code  Site Num  \
0           0           4           13      3002   
1           1           4           13      3002   
2           2           4           13      3002   
3           3           4           13      3002   
4           4           4           13      3002   

                                   Address    State    County     City  \
0  1645 E ROOSEVELT ST-CENTRAL PHOENIX STN  Arizona  Maricopa  Phoenix   
1  1645 E ROOSEVELT ST-CENTRAL PHOENIX STN  Arizona  Maricopa  Phoenix   
2  1645 E ROOSEVELT ST-CENTRAL PHOENIX STN  Arizona  Maricopa  Phoenix   
3  1645 E ROOSEVELT ST-CENTRAL PHOENIX STN  Arizona  Maricopa  Phoenix   
4  1645 E ROOSEVELT ST-CENTRAL PHOENIX STN  Arizona  Maricopa  Phoenix   

   Date Local          NO2 Units  ...          SO2 Units  SO2 Mean  \
0  2000-01-01  Parts per billion  ...  Parts per billion  3.000000   
1  2000-01-01  Parts per billion  ...  Parts per billion  3.000000   
2  2000-01-01  Parts per billion  ...  Parts per billion  2.975000   
3  2000-01-01  Parts per billion  ...  Parts per billion  2.975000   
4  2000-01-02  Parts per billion  ...  Parts per billion  1.958333   

   SO2 1st Max Value  SO2 1st Max Hour SO2 AQI           CO Units   CO Mean  \
0                9.0                21    13.0  Parts per million  1.145833   
1                9.0                21    13.0  Parts per million  0.878947   
2                6.6                23     NaN  Parts per million  1.145833   
3                6.6                23     NaN  Parts per million  0.878947   
4                3.0                22     4.0  Parts per million  0.850000   

   CO 1st Max Value  CO 1st Max Hour CO AQI  
0               4.2               21    NaN  
1               2.2               23   25.0  
2               4.2               21    NaN  
3               2.2               23   25.0  
4               1.6               23    NaN  

[5 rows x 29 columns]

In [34]:
# Here we delete the first column to make it easier to read
df.drop('Unnamed: 0', axis = 1, inplace = True)

In [35]:
df.shape
# We know that there are 28 columns and 1.7 million records.

(1746661, 28)

In [36]:
df.isna().sum()

State Code                0
County Code               0
Site Num                  0
Address                   0
State                     0
County                    0
City                      0
Date Local                0
NO2 Units                 0
NO2 Mean                  0
NO2 1st Max Value         0
NO2 1st Max Hour          0
NO2 AQI                   0
O3 Units                  0
O3 Mean                   0
O3 1st Max Value          0
O3 1st Max Hour           0
O3 AQI                    0
SO2 Units                 0
SO2 Mean                  0
SO2 1st Max Value         0
SO2 1st Max Hour          0
SO2 AQI              872907
CO Units                  0
CO Mean                   0
CO 1st Max Value          0
CO 1st Max Hour           0
CO AQI               873323
dtype: int64

In [37]:
df.describe(include = ['object'])

Address       State       County           City  \
count                  1746661     1746661      1746661        1746661   
unique                     204          47          133            144   
top     PIKE AVE AT RIVER ROAD  California  Los Angeles  Not in a city   
freq                     35332      576142        93381         138411   

        Date Local          NO2 Units           O3 Units          SO2 Units  \
count      1746661            1746661            1746661            1746661   
unique        5996                  1                  1                  1   
top     2002-06-10  Parts per billion  Parts per million  Parts per billion   
freq           640            1746661            1746661            1746661   

                 CO Units  
count             1746661  
unique                  1  
top     Parts per million  
freq              1746661

We know from the above information that there are 47 states, 133 counties, and 144 cities in the dataset. We also know that NO2 and SO2 are expressed in parts per billion while O3 and CO are expressed in parts per million.

In [38]:
df.dtypes

State Code             int64
County Code            int64
Site Num               int64
Address               object
State                 object
County                object
City                  object
Date Local            object
NO2 Units             object
NO2 Mean             float64
NO2 1st Max Value    float64
NO2 1st Max Hour       int64
NO2 AQI                int64
O3 Units              object
O3 Mean              float64
O3 1st Max Value     float64
O3 1st Max Hour        int64
O3 AQI                 int64
SO2 Units             object
SO2 Mean             float64
SO2 1st Max Value    float64
SO2 1st Max Hour       int64
SO2 AQI              float64
CO Units              object
CO Mean              float64
CO 1st Max Value     float64
CO 1st Max Hour        int64
CO AQI               float64
dtype: object

In [39]:
df['Date Local']  = pd.to_datetime(df['Date Local'])

We changed the Date Local column to a datetime from an object (string).

### Graphing

Now we want to find the changes in level of pollution over time vs the APIs above where we were able to glean the current air quality of several states and countries. First, we can find the mean value of pollutants for USA from 2000 to 2016.

In [40]:
# Data
df_AQI_Year = df_AQI[['Date Local', 'NO2 AQI', 'O3 AQI', 'SO2 AQI', 'CO AQI']].resample('Y', on = 'Date Local').mean()
df_AQI_Year.reset_index(inplace=True)

# Plot size
plt.rcParams["figure.figsize"] = (12, 8) 

# Adding plots
fig, ax = plt.subplots()
ax.plot('Date Local', 'NO2 AQI', data=df_AQI_Year, color = "red", label = "NO2 AQI")
ax.plot('Date Local', 'O3 AQI', data=df_AQI_Year, color = "blue", label = "O3 AQI")
ax.plot('Date Local', 'SO2 AQI', data=df_AQI_Year, color = "green", label = "SO2 AQI")
ax.plot('Date Local', 'CO AQI', data=df_AQI_Year, color = "purple", label = "CO AQI")

# Adding plot title and axis titles
plt.title('Mean value of pollutants for USA in 2000-2016', fontsize = 16)
plt.xlabel("Year", fontsize = 13)
plt.ylabel("Mean value of AQI", fontsize = 13)

# change of font size on both axes
ax.tick_params(axis='both', which='major', labelsize=11) 

# Adding legend
ax.legend()
ax.legend(loc='lower center', bbox_to_anchor=(0.5, -.15), ncol = 6, shadow = True, fontsize = 13)

NameError: name 'df_AQI' is not defined

Based on the chart above, we can see that from 2000 to 2016, there has been a steady decline the mean value of the NO2 and SO2 AQI. There has been almost no change in the O3 AQI. 

### Mapping

We can also create some interesting choropleth maps to show geographically the level of pollution over time.

In [ ]:
us_state_abbrev = {
    'Alabama': 'AL',
    'Alaska': 'AK',
    'Arizona': 'AZ',
    'Arkansas': 'AR',
    'California': 'CA',
    'Colorado': 'CO',
    'Connecticut': 'CT',
    'Delaware': 'DE',
    'Florida': 'FL',
    'Georgia': 'GA',
    'Hawaii': 'HI',
    'Idaho': 'ID',
    'Illinois': 'IL',
    'Indiana': 'IN',
    'Iowa': 'IA',
    'Kansas': 'KS',
    'Kentucky': 'KY',
    'Louisiana': 'LA',
    'Maine': 'ME',
    'Maryland': 'MD',
    'Massachusetts': 'MA',
    'Michigan': 'MI',
    'Minnesota': 'MN',
    'Mississippi': 'MS',
    'Missouri': 'MO',
    'Montana': 'MT',
    'Nebraska': 'NE',
    'Nevada': 'NV',
    'New Hampshire': 'NH',
    'New Jersey': 'NJ',
    'New Mexico': 'NM',
    'New York': 'NY',
    'North Carolina': 'NC',
    'North Dakota': 'ND',
    'Ohio': 'OH',
    'Oklahoma': 'OK',
    'Oregon': 'OR',
    'Pennsylvania': 'PA',
    'Rhode Island': 'RI',
    'South Carolina': 'SC',
    'South Dakota': 'SD',
    'Tennessee': 'TN',
    'Texas': 'TX',
    'Utah': 'UT',
    'Vermont': 'VT',
    'Virginia': 'VA',
    'Washington': 'WA',
    'West Virginia': 'WV',
    'Wisconsin': 'WI',
    'Wyoming': 'WY',
}
# To use plotly choropleth maps, states names must be encoded.

In [ ]:
df = df[df['State']!='Country Of Mexico'] # deleting Mexico
df = df[df['State']!='District Of Columbia'] # deleting Columbia
df['State_abbrev'] = df.State.apply(lambda x: us_state_abbrev[x])

Here, we delete the District of Columbia and Mexico because they are not used in the maps.

### Calculating annual mean value of pollutants for every State

In [ ]:

df_AQI = df[['State_abbrev', 'Date Local', 'NO2 AQI', 'O3 AQI', 'SO2 AQI', 'CO AQI']]
df_AQI_State_Year = df_AQI.groupby('State_abbrev').resample('Y', on = 'Date Local').mean()
df_AQI_State_Year.reset_index(inplace = True)
df_AQI_State_Year

In [ ]:
# Adding column corresponding to the year of Date Local column
df_AQI_State_Year['Year'] = df_AQI_State_Year['Date Local'].dt.year
# Sorting values by Date Local (for animated choropleth presented below)
df_AQI_State_Year.sort_values(by = 'Date Local', inplace = True)

### Plotly choropleth showing AQI for Nitrogen Dioxide changes from 2000 to 2016

In [ ]:
fig_NO2 = px.choropleth(df_AQI_State_Year,
              locations = 'State_abbrev',
              animation_frame="Year", # showing changes through the years
              color='NO2 AQI',
              # Creating fixed scale (the same as defined by EPA)
              color_continuous_scale = [(0.00, "green"),   (0.1, "green"),
                                        (0.1, "yellow"), (0.2, "yellow"),
                                        (0.2, "orange"),  (0.3, "orange"),
                                        (0.3, "red"),  (0.4, "red"),
                                        (0.4, "purple"),  (0.6, "purple"),
                                        (0.6, "maroon"),  (1.00, "maroon"),
                                        ],
              range_color = (0, 500),
              locationmode='USA-states',
              scope="usa",
              title='Mean values of Air Quality Index (AQI) per year for Nitrogen Dioxide (NO2)',
              height=600,
             )

# Modifying legend 
fig_NO2.update_layout(coloraxis_colorbar=dict(
    title="Air Quality Index (AQI)",
    ticks="outside", 
    dtick=50
))

### Plotly choropleth showing AQI for Ozone changes from 2000 to 2016

In [ ]:
fig_O3 = px.choropleth(df_AQI_State_Year,
              locations = 'State_abbrev',
              animation_frame="Year", # showing changes through the years
              color='O3 AQI',
              # Creating fixed scale (the same as defined by EPA)
              color_continuous_scale = [(0.00, "green"),   (0.1, "green"),
                                        (0.1, "yellow"), (0.2, "yellow"),
                                        (0.2, "orange"),  (0.3, "orange"),
                                        (0.3, "red"),  (0.4, "red"),
                                        (0.4, "purple"),  (0.6, "purple"),
                                        (0.6, "maroon"),  (1.00, "maroon"),
                                        ],
              range_color = (0, 500),
              locationmode='USA-states',
              scope="usa",
              title='Mean values of Air Quality Index (AQI) per year for Ozone (O3)',
              height=600,
             )

# Modifying legend 
fig_O3.update_layout(coloraxis_colorbar=dict(
    title="Air Quality Index (AQI)",
    ticks="outside", 
    dtick=50
))

### Plotly choropleth showing AQI for Sulfur Dioxide changes from 2000 to 2016

In [ ]:
fig_SO2 = px.choropleth(df_AQI_State_Year,
              locations = 'State_abbrev',
              animation_frame="Year", # showing changes through the years
              color='SO2 AQI',
              # Creating fixed scale (the same as defined by EPA)
              color_continuous_scale = [(0.00, "green"),   (0.1, "green"),
                                        (0.1, "yellow"), (0.2, "yellow"),
                                        (0.2, "orange"),  (0.3, "orange"),
                                        (0.3, "red"),  (0.4, "red"),
                                        (0.4, "purple"),  (0.6, "purple"),
                                        (0.6, "maroon"),  (1.00, "maroon"),
                                        ],
              range_color = (0, 500),
              locationmode='USA-states',
              scope="usa",
              title='Mean values of Air Quality Index (AQI) per year for Sulfur Dioxide (SO2)',
              height=600,
             )

# Modifying legend 
fig_SO2.update_layout(coloraxis_colorbar=dict(
    title="Air Quality Index (AQI)",
    ticks="outside", 
    dtick=50
))


### Plotly choropleth showing AQI for Carbon Monoxide changes from 2000 to 2016

In [ ]:
fig_CO = px.choropleth(df_AQI_State_Year,
              locations = 'State_abbrev',
              animation_frame="Year", # showing changes through the years
              color='CO AQI',
              # Creating fixed scale (the same as defined by EPA)
              color_continuous_scale = [(0.00, "green"),   (0.1, "green"),
                                        (0.1, "yellow"), (0.2, "yellow"),
                                        (0.2, "orange"),  (0.3, "orange"),
                                        (0.3, "red"),  (0.4, "red"),
                                        (0.4, "purple"),  (0.6, "purple"),
                                        (0.6, "maroon"),  (1.00, "maroon"),
                                        ],
              range_color = (0, 500),
              locationmode='USA-states',
              scope="usa",
              title='Mean values of Air Quality Index (AQI) per year for Carbon Monoxide (CO)',
              height=600,
             )

# Modifying legend 
fig_CO.update_layout(coloraxis_colorbar=dict(
    title="Air Quality Index (AQI)",
    ticks="outside", 
    dtick=50
))

As we learned above from the maps, the annual average AQI for all the air pollutants except for O3 has never exceeded 50, which means that the air quality conditions for NO2, SO2 and CO were good (green color).For O3, the AQI has been between 50 and 100 in years 2000-2006 (except 2004) and in 2012.

In [ ]:
# Calculating daily mean value of pollutants for every State
df_AQI_State_Month = df_AQI.groupby('State_abbrev').resample('D', on = 'Date Local').mean()
df_AQI_State_Month.reset_index(inplace = True)

# Adding columns coresponding to the day, month and year of Date Local column
df_AQI_State_Month['Day'] = df_AQI_State_Month['Date Local'].dt.day
df_AQI_State_Month['Month'] = df_AQI_State_Month['Date Local'].dt.month
df_AQI_State_Month['Year'] = df_AQI_State_Month['Date Local'].dt.year

# Data for July in 2005
df_AQI_State_2005_July = df_AQI_State_Month[(df_AQI_State_Month['Year'] == 2005) & (df_AQI_State_Month['Month'] == 7)]

# Sorting values by Date Local (for animated choropleth presented below)
df_AQI_State_2005_July.sort_values(by = 'Date Local', inplace = True)

In [ ]:
# Plotly choropleth showing AQI for Ozone changes for July 2005
fig_O3_v2 = px.choropleth(df_AQI_State_2005_July,
              locations = 'State_abbrev',
              animation_frame="Day", # showing changes through the days
              color='O3 AQI',
              # Creating fixed scale (the same as defined by EPA)
              color_continuous_scale = [(0.00, "green"),   (0.1, "green"),
                                        (0.1, "yellow"), (0.2, "yellow"),
                                        (0.2, "orange"),  (0.3, "orange"),
                                        (0.3, "red"),  (0.4, "red"),
                                        (0.4, "purple"),  (0.6, "purple"),
                                        (0.6, "maroon"),  (1.00, "maroon"),
                                        ],
              range_color = (0, 500),
              locationmode='USA-states',
              scope="usa",
              title='Mean values of Air Quality Index (AQI) per day for Ozone (O3) for July 2005',
              height=600,
             )

# Modifying legend 
fig_O3_v2.update_layout(coloraxis_colorbar=dict(
    title="Air Quality Index (AQI)",
    ticks="outside", 
    dtick=50
))

Here we see that in some cases, the AQI was above 100 and even 150 in July 2005. 

Powerpoint

In [ ]:
pip install python-pptx

In [ ]:
import pptx

In [ ]:
pres = pptx.Presentation()

In [ ]:
len(pres.slides)
# currently 0 slides

In [ ]:
pres.save('air_quality.pptx')

In [ ]:
title_layout = pres.slide_layouts[0]

In [ ]:
slide = pres.slides.add_slide(title_layout)

In [ ]:
blank_slide = pres.slides.add_slide(pres.slide_layouts[6])

In [ ]:
len(pres.slides)

In [ ]:
pres.save('air_quality.pptx')

In [ ]:
len(blank_slide.shapes)

In [ ]:
pres.slides[0].shapes[0].text = 'Air Quality 1'
pres.slides[0].shapes[1].text = 'Air Quality 2'

In [ ]:
pres.save('air_quality.pptx')